node/narwhal/tests/common/primary.rs

```diff
@@ -12,23 +12,193 @@
+                for connection in connections {
+                    debug!("  {}", connection);
+                }
+                sleep(Duration::from_secs(10)).await;
+            }
+        }));
+    }
+}
+
+impl TestNetwork {
+    // Creates a new test network with the given configuration.
+    pub fn new(config: TestNetworkConfig) -> Self {
+        if let Some(log_level) = config.log_level {
+            initialize_logger(log_level);
+        }
+
+        let (accounts, committee) = new_test_committee(config.num_nodes);
+
+        let mut primaries = HashMap::with_capacity(config.num_nodes as usize);
+        for (id, account) in accounts.into_iter().enumerate() {
+            let storage = Storage::new(committee.clone(), MAX_GC_ROUNDS);
+            let ledger = Box::new(MockLedgerService::new());
+            let primary = Primary::<CurrentNetwork>::new(account, storage, ledger, None, Some(id as u16)).unwrap();
+
+            let test_primary = TestPrimary { id: id as u16, primary, sender: None, handles: Default::default() };
+            primaries.insert(id as u16, test_primary);
+        }
+
+        Self { config, primaries }
+    }
+
+    // Starts each node in the network.
+    pub async fn start(&mut self) {
+        for primary in self.primaries.values_mut() {
+            // Setup the channels and start the primary.
+            let (sender, receiver) = init_primary_channels();
+            primary.sender = Some(sender.clone());
+            primary.run(sender.clone(), receiver, None).await.unwrap();
+
+            if self.config.fire_cannons {
+                primary.fire_cannons();
+            }
+
+            if self.config.log_connections {
+                primary.log_connections();
+            }
+        }
+
+        if self.config.connect_all {
+            self.connect_all().await;
+        }
+    }
+
+    // Starts the solution and trasnaction cannons for node.
+    pub fn fire_cannons(&mut self, id: u16) {
+        self.primaries.get_mut(&id).unwrap().fire_cannons();
+    }
+
+    // Connects a node to another node.
+    pub async fn connect(&self, id: u16, peer_id: u16) {
+        let primary = self.primaries.get(&id).unwrap();
+        let peer_ip = self.primaries.get(&peer_id).unwrap().gateway().local_ip();
+        primary.gateway().connect(peer_ip);
+        // Give the connection time to be established.
+        sleep(Duration::from_millis(100)).await;
+    }
+
+    // Connects all nodes to each other.
+    pub async fn connect_all(&self) {
+        for (primary, other_primary) in self.primaries.values().tuple_combinations() {
+            // Connect to the node.
+            let ip = other_primary.gateway().local_ip();
+            primary.gateway().connect(ip);
+            // Give the connection time to be established.
+            tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+        }
+    }
+
+    // Disconnects a node from all other nodes.
```

Comment:
if this disconnects from *all* other nodes, then why the `num_nodes` param? either the comment or the API needs to be adjusted

Reply:
The idea was to make it easy to disconnect N nodes at a time (e.g. 2 in a 4 node network to break quorum). I'll adjust the doc.

Reply:
Happy to go with another setup ofc, just haven't found anything better yet.

Reply:
as long as the docs are adjusted, it should be fine